src/Solvers/multigrid_solver.jl

```diff
@@ -0,0 +1,144 @@
+"""
+    MultigridSolver(grid,
+                    linear_operation!, 
+                    args...;
+                    maximum_iterations = 100,
+                    tolerance = 1e-13,
+                    amg_algorithm = RugeStubenAMG(),
+                    )
+
+Returns a MultigridSolver that solves the linear equation
+``A x = b`` using a multigrid method where `A * x` is
+determined by `linear_operation!`
+
+`linear_operation!` is a function with signature `linear_operation!(Ax, x, args...)` 
+that calculates `A*x` and stores the result in `Ax` for a "candidate solution `x`.
+
+
+The solver is used by calling
+
+```
+solve!(x, solver::MultigridSolver, b; kwargs...)
+```
+
+for `solver`, right-hand side `b`, solution `x`, and optional keyword arguments `kwargs...`.
+
+Arguments
+=========
+
+* `maximum_iterations`: Maximum number of iterations the solver may perform before exiting.
+
+* `tolerance`: Tolerance for convergence of the algorithm. The algorithm quits when
+               `norm(A * x - b) < tolerance`.
+"""
+function MultigridSolver(grid,
+                        linear_operation!, 
+                        args...;
+                        maximum_iterations = 100, #prod(size(template_field)),
+                        tolerance = 1e-13, #sqrt(eps(eltype(template_field.grid))),
+                        amg_algorithm = RugeStubenAMG(),
+                        )
+
+    arch = architecture(grid)
+
+    A = create_matrix(grid, linear_operation!, args...)
+
+    return MultigridSolver(arch,
+                           grid,
+                           A,
+                           tolerance,
+                           maximum_iterations,
+                           amg_algorithm
+                           )
+end
+
+# TODO make inplace create_matrix!
+
+# For free surface without Δt
+function create_matrix(grid, linear_operator!, ::Any , ::Any, ::Any, ::Nothing)
+    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
+    return spzeros(eltype(grid), Nx*Ny, Nx*Ny)
+end
+
+function create_matrix(grid, linear_operator!, args...)
+    Nx, Ny = grid.Nx, grid.Ny
+    A = spzeros(eltype(grid), Nx*Ny, Nx*Ny)
+
+    make_column(f) = reshape(interior(f), (Nx*Ny, 1))
+
+    eᵢⱼₖ = Field{Center, Center, Nothing}(grid)
+    ∇²eᵢⱼₖ = Field{Center, Center, Nothing}(grid)
+    for j in 1:Ny, i in 1:Nx
+        eᵢⱼₖ .= 0
+        ∇²eᵢⱼₖ .= 0
+        eᵢⱼₖ[i, j] = 1
+        fill_halo_regions!(eᵢⱼₖ)
+        linear_operator!(∇²eᵢⱼₖ, eᵢⱼₖ, args...)
+
+        A[:, Nx*(j-1)+i] = make_column(∇²eᵢⱼₖ)
+    end
+    
+    return A
+end
+
+"""
+    solve!(x, solver::MultigridSolver, b, args...)
+
+Solve `A * x = b` using a multigrid method, where `A * x` is
+determined by `linear_operation!` given in the MultigridSolver constructor.
+"""
+
+function solve!(x, solver::MultigridSolver, b; kwargs...)
+    grid = b.grid
+    Nx, Ny, Nz = size(grid)
+    b_array = collect(reshape(interior(b), Nx * Ny * Nz))
+    x_array = collect(reshape(interior(x), Nx * Ny * Nz)) 
```

Comment:
I'm wondering whether these two should be fields of `solver` and this way we don't need to allocate them here